### ISSUE TITLE
On/Off Switch Updated to True

### ISSUE THREAD

**Issue by emst00** [lukasroegner/homebridge-apple-tv-remote#30]
Thank you for this plug-in, I find it very helpful. My switch gets stuck to "on" the majority of the time when the AppleTV is off and can't figure out why. I enabled debug mode and this is what I get:

[4/25/2020, 9:02:05 PM] [Apple TV Platform] [AppleTV] Sending heartbeat...
[4/25/2020, 9:02:05 PM] [Apple TV Platform] [AppleTV] Connecting...
[4/25/2020, 9:02:05 PM] [Apple TV Platform] [AppleTV] Force reconnect: false
[4/25/2020, 9:02:05 PM] [Apple TV Platform] [AppleTV] Retry count: 10
[4/25/2020, 9:02:05 PM] [Apple TV Platform] [AppleTV] Already connected
[4/25/2020, 9:02:05 PM] [Apple TV Platform] [AppleTV] On/off switch updated to true

**Comment by Kienz** [comment]
I have the same issue - the problem was, that the Apple TV (Home Hub) was used to play music to Airplay Speakers. After that the ```logicalDeviceCount``` was 1 even though the Apple TV was shutdown.
In this case ```isGroupLeader```and ```isAirplayActive``` are true.
```
DeviceInfo DeviceInfoMessage {
  airplayReceivers: [ 'TVAirPlay' ],
  uniqueIdentifier: 'xxx',
  name: 'Apple TV Schlafzimmer',
  localizedModelName: 'Apple TV',
  systemBuildVersion: '17L256',
  applicationBundleIdentifier: 'com.apple.mediaremoted',
  protocolVersion: 1,
  lastSupportedMessageType: 78,
  supportsSystemPairing: true,
  allowsPairing: true,
  systemMediaApplication: 'com.apple.TVMusic',
  supportsACL: true,
  supportsSharedQueue: true,
  supportsExtendedMotion: true,
  sharedQueueVersion: 2,
  deviceUID: 'xxx',
  managedConfigDeviceID: 'xxx',
  deviceClass: 4,
  logicalDeviceCount: 1,
  isProxyGroupPlayer: true,
  groupUID: 'xxx',
  isGroupLeader: true,
  isAirplayActive: false,
  systemPodcastApplication: 'com.apple.podcasts',
  enderDefaultGroupUID: 'xxx'
}
```

**Comment by lukasroegner** [comment]
@Kienz Those are great findings! Have you figured out a way to determine whether the Apple TV is off despite `logicalDeviceCount: 1`?

**Comment by Kienz** [comment]
I've changed the `currentPowerState >= 1` to the following.
```
var currentPowerState = deviceInfo.payload.logicalDeviceCount,
    proxyGroupPlayer = deviceInfo.payload.isProxyGroupPlayer,
    airplayActive = deviceInfo.payload.isAirplayActive;

if (
    (currentPowerState >= 1 && !proxyGroupPlayer) ||
    (currentPowerState >= 1 && proxyGroupPlayer && airplayActive)  
) {
    that.updatePowerState(true);
} else {
```

Have to check the different variants. With this code the Apple TV is displayed Off. But have to check what happens when AirPlay to a Speaker over HomeKit is started.

**Comment by lukasroegner** [comment]
@Kienz Have you already checked whether your solutions works for all cases?

**Comment by Kienz** [comment]
I think everything works - but not sure if I tested all cases. :#

**Comment by lukasroegner** [comment]
I integrated your checks into the plugin and published a beta version:
```
npm install -g homebridge-apple-tv-remote@beta
```

**Comment by lukasroegner** [comment]
Has anyone had issues with the beta version? Does it enhance the on/off detection for you?

**Comment by RaymondMouthaan** [comment]
Hi @lukasroegner, just installed the beta, but t seems to be versioned as `homebridge-apple-tv-remote v0.7.7`? 

<img width="735" alt="image" src="https://user-images.githubusercontent.com/25717738/80909547-7f787e00-8d29-11ea-89fa-ad685c57a1c6.png">

My config is like:

```
{
            "name": "Apple TV Platform",
            "devices": [
                {
                    "name": "AppleTV",
                    "credentials": "token",
                    "isPlayPauseSwitchEnabled": true,
                    "playPauseSwitchName": "playPause"
                }
            ],
            "platform": "AppleTvPlatform"
        }
```

I've only configured a PlayPause switch since im interested to control some lights based upon its state. (on play -> lights off, pause lights on). After some testing with a Harmony remote, which play/pauses the Apple TV, the PlayPause switch is not always in sync with the actual status of the AppleTV. The switch even changes its state while doing nothing. Like now a movie is playing, the PlayPause switch has state `on`, but after a few minutes it changed to state `off` and while writing this it changed back to state 'on' again ... and now its 'off' again ...

How do I enable debugging, so I can provide the logs?

**Comment by lukasroegner** [comment]
@RaymondMouthaan see #7. There are several issues when detecting Play/Pause state:
* Inconsistent implementation in tvOS apps (e.g. Netflix auto play)
* Delays with updates of the play/pause state (in my case, Amazon Prime Video seems to have some issues)
* No play/pause state at all (e.g. German news app "Welt" does not report state for their live stream)

You can see the raw message coming from the Apple TV with the `appletv` command line tool. If you recognize a pattern, which can enhance the play/pause state, feel free to send me your findings.

**Comment by sergoo-lbt** [comment]
> Has anyone had issues with the beta version? Does it enhance the on/off detection for you?

I tried the beta version. My power switch does not behave adequately at all. I think the latest version is more stable. Sometimes the status stops changing and I restart homebridge, after that everything works perfectly again for a while.

**Comment by lukasroegner** [comment]
@RaymondMouthaan I published a new beta version (0.8.0) on NPM. Changes:
* New playback state detection (same as the `appletv state` command)
* New debug mechanism (start homebridge with parameter `-D` to get all messages received by the plugin in the log output)

**Comment by RaymondMouthaan** [comment]
@lukasroegner just for reference https://github.com/lukasroegner/homebridge-apple-tv-remote/issues/7#issuecomment-623089316.

Just installed beta 0.8.0 and enabled debug option -D to homebridge, this gives enormous amount of information, which is to much to handle ...

However I disabled the debug logging and it looks much more stable then before! The homebridge-apple-tv-remote PlayPause switch seems to follow the actual AppleTV device state as it suppose to do, which is awesome! 

Tested also with Prime Video, which is shows exact behaviour, so it looks like you smashed this bug 💯

**Comment by lukasroegner** [comment]
@RaymondMouthaan That's great! If your testing is completed, I'll integrate the beta branch into master. Can you also check whether the on/off state is now working properly for you?

**Comment by RaymondMouthaan** [comment]
@lukasroegner as for now PlayPause state is working perfectly fine and stable 👍 

Combined with OnOff switch it looks like the OnOff switch perfectly follows the actual state of the Apple TV device. However one thing I noticed is that PlayPause doesn't turn off when the device is turned off, while is should -  after all, it can never be played when the device is turned off.

**Comment by lukasroegner** [comment]
@RaymondMouthaan New beta version 0.8.1: When the plugin detects that the state of the Apple TV switched to off, the play state is also switched to off. Please check if that works properly for you

**Comment by RaymondMouthaan** [comment]
@lukasroegner wow that was quick .. Im on it 👍

**Comment by sergoo-lbt** [comment]
> @RaymondMouthaan New beta version 0.8.1: When the plugin detects that the state of the Apple TV switched to off, the play state is also switched to off. Please check if that works properly for you

You are an amazing developer! Thank you very much for your work. I'll test everything today!

**Comment by sergoo-lbt** [comment]
> @RaymondMouthaan New beta version 0.8.1: When the plugin detects that the state of the Apple TV switched to off, the play state is also switched to off. Please check if that works properly for you

There's still something wrong. Apple tv itself went to sleep, but the power and pause button status remained on. Turned off the TV, it turns off the apple tv via hdmi cec, but in the home app the power switch and pause are all also enabled.

**Comment by lukasroegner** [comment]
@sergoo-lbt Please start homebridge with `-D` parameter and let the Apple TV go to sleep again. During waiting, you should get at least one (large) log entry per minute. I can evaluate the logs to see what data is sent by the Apple TV when it goes to sleep.

**Comment by sergoo-lbt** [comment]
> @sergoo-lbt Please start homebridge with `-D` parameter and let the Apple TV go to sleep again. During waiting, you should get at least one (large) log entry per minute. I can evaluate the logs to see what data is sent by the Apple TV when it goes to sleep.
Here's what I was able to understand. Basically, everything works. But it takes a very long time to change the power state and pause, about a minute. I guess I just didn't wait for them to be fully updated. It turns out that if you turn on the power or pause key several times in a row, the status will not have time to update. I'll try again. I just shared my thoughts. I'm trying to figure out how it all works. I'll attach the logs, but I apologize for the screenshots, I just don't have access to my computer right now.
![EC513861-652A-40BA-ADAF-2085982C08A9](https://user-images.githubusercontent.com/64399012/80924092-88f0fd00-8da0-11ea-8ef6-9907f73f518b.png)
![65356B42-7B2C-43A7-9439-EB8B9D55C378](https://user-images.githubusercontent.com/64399012/80924102-90b0a180-8da0-11ea-8bcc-d607b329b59e.png)
![C3D21C70-4298-4C61-B286-19DDB13CD66F](https://user-images.githubusercontent.com/64399012/80924105-94dcbf00-8da0-11ea-874a-1061c879ba58.png)
![91DF2089-88DD-4935-80DE-DE8B7083D8F8](https://user-images.githubusercontent.com/64399012/80924109-973f1900-8da0-11ea-86fa-65c31087eaaa.png)

**Comment by sergoo-lbt** [comment]
When I took screenshots, the apple tv was turned off.

**Comment by lukasroegner** [comment]
So, I did another update of the beta version (0.8.2):
* On/off state is actively requested from the Apple TV when the play state changes (enhances delay of the on/off state update in some cases)
* Enhanced the debug output, so that it only shows important information

I also tested this version extensively, here are my findings:
* The play/pause state now works fine and updated very quickly
* Both tests of @RaymondMouthaan worked for me (movie playback -> turn Apple TV off via remote -> play state is off)
* The on/off state sometimes instantly updated in HomeKit, sometimes it takes longer to update

Background info: The Apple TV should send messages to the plugin whenever the device state changes (important for detecting on/off state). This, however, does not seem to work reliable. Thus, I implemented a "heartbeat" some time ago, which actively requests the current device state from the Apple TV (this works fine). The current interval is 60 seconds (i.e. every minute the on/off state is updated reliably. You can change the interval with an undocumented config property: in the root of the plugin configuration, add `"heartbeatInterval": 10` for a 10 second interval. This can increase the responsiveness, however, will also increase network traffic.

@sergoo-lbt From your logs, I assume that you also have the issue that the on/off state is not updated instantly when turning off the Apple TV. Try (a) the new version, which requests device info update when the play state changes, and (b) to decrease the heartbeat interval.

**Comment by RaymondMouthaan** [comment]
@lukasroegner, thanks for a new beta. Just updated to this beta (0.8.2) and I'll test it and keep you informed. Good to know about this undocumented `heatbeatInterval`, i'll leave it as it, but might change it if some weird behaviour occurs.

**Comment by sergoo-lbt** [comment]
@lukasroegner, I did as you said. Added a heartbeat, 10 seconds. I'll attach the result. I did not wait for the change in the power status. With a pause, everything seems to be working fine!
![738E51A1-DE7A-4A3A-97CA-C59A50022B99](https://user-images.githubusercontent.com/64399012/80925321-47fce680-8da8-11ea-8c77-5ce072309f22.png)
![8C3744D2-D1C7-4CB6-B2D6-5CE786EC2D39](https://user-images.githubusercontent.com/64399012/80925322-4cc19a80-8da8-11ea-9ff0-2fc105f8773c.png)
![169D6949-6D8C-42A3-A7C5-994DCE5E261B](https://user-images.githubusercontent.com/64399012/80925323-4d5a3100-8da8-11ea-84ab-dc597f699604.png)
![B5566395-9A8E-4441-B4AA-8673B962D45F](https://user-images.githubusercontent.com/64399012/80925325-4f23f480-8da8-11ea-9543-e8bf0eca2118.png)

**Comment by sergoo-lbt** [comment]
@lukasroegner ,  I still waited. He updated the status himself. But it was really long, 12 minutes...
![8600CA53-4FA4-4593-812F-B386C6022C39](https://user-images.githubusercontent.com/64399012/80925413-fb65db00-8da8-11ea-98b1-4860322ea1a3.png)

**Comment by sergoo-lbt** [comment]
@lukasroegner , What is the minimum value of "heartbeatInterval"?

**Comment by sergoo-lbt** [comment]
@lukasroegner , Apple tv has been turned off for 10 minutes. The log is always the same.
![8DCF77C8-DF80-48C1-9DCF-E223CBEC7EFF](https://user-images.githubusercontent.com/64399012/81013257-4698f000-8e74-11ea-843e-b981e1e5dd27.jpeg)

**Comment by lukasroegner** [comment]
@sergoo-lbt Even with smaller heartbeat interval, the plugin does not get the Apple TV responses back properly (can be seen from the logs). I would suggest to do a restart of the Apple TV, it seems that there is something wrong.

**Comment by sergoo-lbt** [comment]
Rebooted apple tv. Not help. Only restarting homebridge temporarily helps.

**Comment by lukasroegner** [comment]
@sergoo-lbt Have you tried to change the setup (different Homebridge host, Wi-Fi vs. cable) to determine the cause of your issue?

> Only restarting homebridge temporarily helps.

Does that mean that the delay is not present when starting up Homebridge and occurs after some time?

**Comment by sergoo-lbt** [comment]
> @sergoo-lbt Have you tried to change the setup (different Homebridge host, Wi-Fi vs. cable) to determine the cause of your issue?
> 

You can't connect the cable. I'll experiment with the host! Thanks!

> > Only restarting homebridge temporarily helps.
> 
> Does that mean that the delay is not present when starting up Homebridge and occurs after some time?

Yes, that's right! The status is still updated, but after a very long period of time.

**Comment by RaymondMouthaan** [comment]
@lukasroegner, here all seems to work stable now .. as for me you can release the beta :-)

**Comment by lukasroegner** [comment]
Beta is now merged into master, new version 0.8.3 on NPM.

**Comment by RaymondMouthaan** [comment]
> Beta is now merged into master, new version 0.8.3 on NPM.

Awesome! 👍
